### DISCUSSION TITLE
Upgrade path for SecureStorage when moving from Xamarin.Essentials to Maui.Essentials

### DISCUSSION THREAD

**Discussion by queequac** [dotnet/maui#10226] [Q&A] [answered]
Is there an upgrade path from `Xamarin.Essentials` to `Maui.Essentials` when using the `SecureStorage`?

I stored data in my app via Xamarin in the past, after porting to net6.0-android I'd use maui.essentials.
But seems that I can no longer read data via Maui which was previously stored with Xamarin. (Seems there changed more on encryption that just the `Alias` name.)

Is there any upgrade path?
Or do I have to ask my app users again for all the stuff I stored safely?😕

**Comment by nicjay** [comment, marked as answer]
Xamarin and MAUI encrypt the SecureStorage contents differently. But it does seem to be possible to bring in the Xamarin.Essentials package, retrieve the values (SecureStorage.GetAsync), then re-save them with MAUI Essentials. 

I found this reddit thread with more discussion and code samples: https://www.reddit.com/r/dotnetMAUI/comments/wswyyx/complete_rewrite_of_my_xamarin_app_to_net_maui/

**Reply by queequac** [reply]
Using Xamarin.Essentials for reading and then storing in Maui.Essentials was not really a solution for me, since I'd have to reference both packages then for quite a while - you never know when people actually upgrade. Then I could even stick to Xamarin.Essentials only.
Similar to the reddit post, I extracted the old SecureStorage code and added it to my code base. But I will most likely keep on using that one only, since I need that code for upgrading now anyway, and the new SecureStorage seems to require a brief moment of reflection at least on the very first access - and I don't want to slow done my startup performance😉.

I am quite disappointed that Maui.Essentials provide the same API for an easy transition, but the actual implementation allows only developers to easily upgrade, not the users of our apps. I'd have expected that the lecacy securestorage implementation would have been in Maui.Essentials in addition to the (new) default one, so people can actually keep on reading their data without workarrounds.

Anyway, seems there is indeed nothing taken into consideration by the library, so I will mark the question as answered.

**Comment by jfversluis** [comment]
Hey there! I have created a library that should help with this scenario. You can find it [here](https://github.com/jfversluis/Plugin.Maui.FormsMigration) (and on NuGet).

Additionally you can find guidance for this scenario in the Docs, [here](https://learn.microsoft.com/dotnet/maui/migration/secure-storage).

I'd be curious to hear your findings!
